arangosh/Backup/BackupFeature.cpp

```diff
@@ -0,0 +1,509 @@
+                                double originalUptime, double maxWaitForRestart) {
+  arangodb::Result result;
+
+  auto start = std::chrono::high_resolution_clock::now();
+  auto timeSinceStart = [start]() -> double {
+    auto now = std::chrono::high_resolution_clock::now();
+    auto duration = now - start;
+    return (static_cast<double>(duration.count()) /
+            decltype(duration)::period::den* decltype(duration)::period::num);
+  };
+
+  LOG_TOPIC(INFO, arangodb::Logger::BACKUP)
+      << "Waiting for server to restart...";
+
+  // sleep once to allow shutdown to start
+  std::this_thread::sleep_for(std::chrono::seconds(1));
+
+  while (timeSinceStart() < maxWaitForRestart) {
+    std::unique_ptr<arangodb::httpclient::SimpleHttpClient> client;
+    try {
+      result = clientManager.getConnectedClient(client, true, false, false, true);
+      if (result.ok() && client != nullptr) {
+        double uptime = 0.0;
+        result = ::getUptime(*client, uptime);
+        if (result.ok() && uptime < originalUptime) {
+          LOG_TOPIC(INFO, arangodb::Logger::BACKUP)
+              << "...server back up and running!";
+          return result;
+        }
+        // server still down, or still up prior to restart; fall through, wait, retry
+      }
+    } catch (...) {
+    }  // ignore errors here
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+  }
+
+  // okay, we timed out
+  result.reset(
+      TRI_ERROR_HTTP_SERVICE_UNAVAILABLE,
+      "Server failed to respond to requests in the expected timeframe.");
+
+  return result;
+}
+
+arangodb::Result executeList(arangodb::httpclient::SimpleHttpClient& client,
+                             arangodb::BackupFeature::Options const& options) {
+  arangodb::Result result;
+
+  std::string const url = "/_admin/hotbackup/list";
+  std::unique_ptr<arangodb::httpclient::SimpleHttpResult> response(
+      client.request(arangodb::rest::RequestType::POST, url, nullptr, 0));
+  auto check = ::checkHttpResponse(client, response);
+  if (check.fail()) {
+    return check;
+  }
+
+  // extract vpack body from response
+  std::shared_ptr<VPackBuilder> parsedBody;
+  try {
+    parsedBody = response->getBodyVelocyPack();
+  } catch (...) {
+    result.reset(::ErrorMalformedJsonResponse);
+    return result;
+  }
+  VPackSlice const resBody = parsedBody->slice();
+  VPackSlice const resultObject = resBody.get("result");
+  TRI_ASSERT(resultObject.isObject());
+  VPackSlice const backups = resultObject.get("hotbackups");
+  TRI_ASSERT(backups.isArray());
+
+  if (backups.isEmptyArray()) {
+    LOG_TOPIC(INFO, arangodb::Logger::BACKUP)
+        << "There are no backups available.";
+  } else {
+    LOG_TOPIC(INFO, arangodb::Logger::BACKUP)
+        << "The following backups are available:";
+    for (auto const& backup : VPackArrayIterator(backups)) {
+      TRI_ASSERT(backup.isString());
+      LOG_TOPIC(INFO, arangodb::Logger::BACKUP) << " - '" << backup.copyString() << "'";
```

Comment:
is this properly cut'n'pasteable?

Reply:
You can copy the portion between the `'`s, yes. Should we omit the `'`s and just let it be whitespace around?

Reply:
idk whether other terminals don't do this right, xterm can, does putty?

Reply:
I have no idea. There's an issue with copying and pasting strings in different terminals?